Masq 191 hash keys

leventdem opened the pull request:
We want to hash the keys with a nonce: 
Each call to put/get with key, nonce => hash(nonce || key)

Commit:
add encodingFormat and export hash256 method

Commit:
add hashKey method, update put/get which take nonce as input, update list

Commit:
update test

Commit:
add an error

Commit:
add encoding format to genRandomBuffer, add _checkEncodingFormat function

Commit:
update test of genRandomBuffer

Commit:
small refactoring

Commit:
add nonce generation alongside the masterkey

Commit:
add nonce generation, decryptMasterKey return not only the master key, but also the nonce

Commit:
small improve

sujeebant commented on src/errors.js:
```diff
@@ -1,4 +1,5 @@
 const ERRORS = {
+  INVALID_ENCODING_FORMAT: 'Supporting encoding format: hex and base64',
```
Supported encoding formats: hex and base64
OR Supporting encoding formats hex and base64

sujeebant commented on src/utils.js:
```diff
@@ -44,20 +44,38 @@ function dbExists (dbName) {
    * @returns {Promise}
    */
-const get = async (db, encKey, key) => {
+const get = async (db, encKey, key, nonce) => {
```
You could reorder the parameters to have nonce after encKey

sujeebant commented on src/utils.js:
```diff
@@ -44,20 +44,38 @@ function dbExists (dbName) {
-const get = async (db, encKey, key) => {
+const get = async (db, encKey, key, nonce) => {
   if (!db) throw new MasqError(ERRORS.NO_DB)
   if (!encKey) throw new MasqError(ERRORS.NO_ENCRYPTION_KEY)
```
Check if nonce argument is given

sujeebant commented on src/utils.js:
```diff
@@ -86,21 +112,24 @@ const list = async (db, encKey, prefix) => {
+    return {
+      key: dec.key,
+      value: dec.value
+    }
```
Can't you return dec directly ?

G-Ray replied:
does dec contains more values than just `key` and `value` ?

sujeebant commented on src/utils.js:
```diff
@@ -86,21 +112,24 @@ const list = async (db, encKey, prefix) => {
+  }))
 
   const reformattedDic = decList.reduce((dic, e) => {
-    const el = Array.isArray(e) ? e[0] : e
```
:+1:

Commit:
remove first / or last / in key name for put method

Commit:
change name (add Nonce) and some descriptions

Commit:
add no nonce error and fix typo

Commit:
Refactor

- Check given key for hashKey
- Re-order parameteres of put/get
- Add nonce to list (in case of a prefix we need to hash it)

Commit:
rename and refactor

Commit:
change parameters order for put/get, update test

Commit:
add watcher

Commit:
small improvements, return dec directly

Commit:
add a generic hashMsg method

G-Ray commented on src/crypto.js:
```diff
@@ -1,8 +1,15 @@
+  if (format !== 'hex' && format !== 'base64') throw new MasqError(ERRORS.INVALID_ENCODING_FORMAT)
+}
+
+const genRandomBuffer = (len = 16, encodingFormat) => {
```
The return output is inconsistent: it either return a Buffer, or a string.

I think we should keep this function simple by just generating a random buffer. The user of this function will call `toString(encoding)` if we wants

G-Ray commented on src/utils.js:
```diff
@@ -44,71 +44,125 @@ function dbExists (dbName) {
+ */
+const hashKey = async (key, nonce) => {
+  if (key === '/' || key === '') return '/'
+  const prefixes = key.split('/').filter(prefix => prefix !== '')
```
ok

sujeebant replied:
Why filter all empty strings ? is a key like '/abc//def' valid ?

sujeebant commented on src/utils.js:
```diff
@@ -44,71 +44,125 @@ function dbExists (dbName) {
+
+  if (sanitizedKey[sanitizedKey.length - 1] === '/') {
+    sanitizedKey = sanitizedKey.slice(0, -1)
+  }
```
Add a comment to explain that the key stored in the value is sanitized to remove the '/' at the beginning and end of the '/' separated key.

G-Ray replied:
~~IIUC this sanitization step is not needed, as `hashKey()` will remove unnecessary slashes anyway ?~~

G-Ray commented on src/utils.js:
```diff
@@ -44,71 +44,125 @@ function dbExists (dbName) {
+  if (key === '/' || key === '') return '/'
+  const prefixes = key.split('/').filter(prefix => prefix !== '')
+  const hashedPrefixes = await Promise.all(prefixes.map(async (prefix) => hash256(`${nonce}${prefix}`)))
+  return hashedPrefixes.join('/')
```
this will miss the leading / right ? is it an issue ?

sujeebant commented on src/utils.js:
```diff
@@ -44,71 +44,125 @@ function dbExists (dbName) {
-  })))
+  const decList = await Promise.all(list.map(async (elt) => {
+    const dec = await decrypt(encKey, elt.value)
+    return dec
```
I think we can just use await Promise.all(list.map(elt => decrypt(encKey, elt.value)))

G-Ray commented on test/utils.js:
```diff
@@ -28,6 +29,14 @@ describe('MasqCommon utils', () => {
+    it('Should hash each subpaths of a given key', async () => {
+      const key = '/favoris/paris'
+      const prefixes = await MasqCommon.utils.hashKey(key, nonce)
+      const expected = 'ab595ffb49ed89da9e15a501d65f2fe0d0c47ef41ff346c2cd31246dd5c54003/e13e07a3ac6a83421551448b2ba94205a0fdaca27d622addc8bab44300f28290'
```
I think we should keep the leading '\', what to you think ?

G-Ray commented on test/crypto.js:
```diff
@@ -2,215 +2,233 @@
-      const sameMK = await MasqCommon.crypto.decryptMasterKey(passphrase, protectedMK)
-      const sameCryptokey = await MasqCommon.crypto.importKey(sameMK)
-      const dec = await MasqCommon.crypto.decrypt(sameCryptokey, enc)
+      const sameMKAndNonce = await crypto.decryptMasterKeyAndNonce(passphrase, protectedMK)
```
don't hesitate to do `const { masterKey } = await...`

G-Ray commented on test/crypto.js:
```diff
@@ -2,215 +2,233 @@
 /* global chai */
 
+const { ERRORS } = MasqCommon.errors
+const { crypto } = MasqCommon
```
+1

G-Ray replied:
if you have other problems with globals, you should be able to declare these var locally inside the describe section

G-Ray commented on test/utils.js:
```diff
@@ -39,37 +48,73 @@ describe('MasqCommon utils', () => {
+      const db = await MasqCommon.utils.createPromisifiedHyperDB('dB9')
+      await MasqCommon.utils.dbReady(db)
+      await MasqCommon.utils.put(db, masterKey, nonce, `${prefix1}one`, item)
+      const list1 = await MasqCommon.utils.list(db, masterKey, nonce, prefix1)
```
maybe we can verify that the retrieved item in the list match the exepected one ?

Commit:
add genRandomBufferAsStr

Commit:
remove unnecessary async await in promise.all

Commit:
update test with a deepEqual instead of testing the object keys

Commit:
remove log

G-Ray commented on src/crypto.js:
```diff
@@ -1,10 +1,22 @@
+  if (encodingFormat) {
+    _checkEncodingFormat(encodingFormat)
+  }
+  const values = window.crypto.getRandomValues(new Uint8Array(len))
```
you should reuse your function `genRandomBuffer` here :)

leventdem replied:
yes, right.

G-Ray commented on src/crypto.js:
```diff
@@ -1,10 +1,22 @@
+    _checkEncodingFormat(encodingFormat)
+  }
+  const values = window.crypto.getRandomValues(new Uint8Array(len))
+  return encodingFormat ? Buffer.from(values).toString(encodingFormat) : Buffer.from(values)
```
Once you get the buffer with `genRandomBuffer`, return a string, an use a default encoding if none is specified

Commit:
small fixes

Commit:
remove not necessary Buffer.from

G-Ray commented on src/crypto.js:
```diff
@@ -1,10 +1,22 @@
+    _checkEncodingFormat(encodingFormat)
+  }
+  const buf = genRandomBuffer(len)
+  return buf.toString(encodingFormat)
```
:+1:

G-Ray commented:
We should set the key, dek and nonce only once, to avoid passing it around each function `get` `put` etc...

Commit:
add a test for hashKey with another nonce and renaming

G-Ray commented on test/utils.js:
```diff
@@ -19,109 +24,182 @@ describe('MasqCommon utils', () => {
-      await MasqCommon.utils.put(db, masterKey, 'one', item)
-      const dec = await MasqCommon.utils.get(db, masterKey, 'one')
+      const keyName = 'one'
+      // need to check if encryption is ok
```
need to check ?

leventdem replied:
no old comment

Commit:
remove log

Commit:
add a test to check if two consecutive '/' is corretly handled
